pkg/rc/replication_controller.go

```diff
@@ -633,81 +669,276 @@ func (rc *replicationController) unschedule(txn *auditingTransaction, node types
+func (rc *replicationController) transferNodes(ineligible []types.NodeName) error {
+	if rc.isWatchingHealth {
+		// a node transfer to replace the ineligible node is already in progress
+		return nil
 	}
-	if err != nil {
-		return util.Errorf("Unable to transfer nodes: %s", err)
+
+	if rc.AllocationStrategy != fields.CattleStrategy {
+		errMsg := fmt.Sprintf("Non-cattle RC has scheduled %d ineligible nodes: %s", len(ineligible), ineligible)
+		err := rc.alerter.Alert(rc.alertInfo(errMsg), alerting.LowUrgency)
+		if err != nil {
+			rc.logger.WithError(err).Errorln("Unable to send alert")
+		}
+		// We don't return an error here because this is the current behavior for non-cattle
+		// ineligible nodes: only page
+		return nil
+	}
+
+	status, _, err := rc.rcStatusStore.Get(rc.ID())
+	if err != nil && !statusstore.IsNoStatus(err) {
+		return err
+	}
+
+	var newNode, oldNode types.NodeName
+	if status.NodeTransfer == nil {
+		newNode, oldNode, err = rc.updateAllocations(ineligible)
+		if err != nil {
+			return err
+		}
+
+		err = rc.scheduleWithSession(newNode)
+		if err != nil {
+			// Different RC may have already taken over the new node, abort transfer
+			deleteErr := rc.rcStatusStore.Delete(rc.ID())
+			if deleteErr != nil {
+				rc.logger.WithError(deleteErr).Errorln("could not delete rc status")
+			}
+			return util.Errorf("Could not schedule with session; new rc may have taken node: %s", err)
+		}
+	} else {
+		newNode = status.NodeTransfer.NewNode
+		oldNode = status.NodeTransfer.OldNode
 	}
 
+	rc.nodeTransfer.newNode = newNode
+	rc.nodeTransfer.oldNode = oldNode
+	rc.nodeTransfer.quit = make(chan struct{})
+
+	go rc.doBackgroundNodeTransfer()
+
 	return nil
 }
 
-func (rc *replicationController) transferCattleNodes(ineligibleCurrent []types.NodeName) error {
-	if len(ineligibleCurrent) < 1 {
-		return util.Errorf("Need at least one ineligible node to transfer from")
+func (rc *replicationController) updateAllocations(ineligible []types.NodeName) (types.NodeName, types.NodeName, error) {
+	if len(ineligible) < 1 {
+		return "", "", util.Errorf("Need at least one ineligible node to transfer from, had 0")
 	}
 
-	status, queryMeta, statusErr := rc.rcStatusStore.Get(rc.ID())
-	if statusErr != nil && !statusstore.IsNoStatus(statusErr) {
-		return util.Errorf("Unable to get rc status: %s", statusErr)
-	}
+	rc.mu.Lock()
+	man := rc.Manifest
+	sel := rc.NodeSelector
+	rc.mu.Unlock()
 
-	if status.NodeTransfer != nil {
-		// Node transfer already in progress
-		return nil
+	oldNode := ineligible[0]
+	err := rc.scheduler.DeallocateNodes(sel, []types.NodeName{oldNode})
+	if err != nil {
+		return "", "", util.Errorf("Could not deallocate from %s: %s", oldNode, err)
 	}
 
 	nodesRequested := 1 // We only support one node transfer at a time right now
-	rc.mu.Lock()
-	newNodes, err := rc.scheduler.AllocateNodes(rc.Manifest, rc.NodeSelector, nodesRequested)
-	rc.mu.Unlock()
+	newNodes, err := rc.scheduler.AllocateNodes(man, sel, nodesRequested)
```

Comment:
I'm wondering if this spot should check if there already is a newly allocated node. I'm not sure how that would happen, I guess if the allocator server makes the allocation on a previous iteration but we end up with an error.

You could feed the diff of current and eligible down to here to see if there is already an unused allocation to use for the transfer rather than calling indiscriminately.

I'm not sure how important it is to check for this, I think it might mess up the rolling update logic if there's an extra allocation lying around unused